### PR TITLE
Use generics for ordered map, lazily allocate inner map and list

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Use proper Go generics for the ordered map implementation
- Remove all the generated implementations 
- Optimize the allocations of the inner map and list by making them lazy

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/common/orderedmap/orderedmap.go [hunk 1]
```diff
@@ -108,23 +121,39 @@ func (om *KeyTypeValueTypeOrderedMap) Delete(key KeyType) (oldValue ValueType, p
 }
 
 // Len returns the length of the ordered map.
-func (om *KeyTypeValueTypeOrderedMap) Len() int {
+func (om OrderedMap[K, V]) Len() int {
+	if om.pairs == nil {
+		return 0
+	}
+
```
Comment (SupunS) [inline runtime/common/orderedmap/orderedmap.go:128]: I think builtin `len()` function already handles `nil`. But no harm having it here I guess

File: runtime/sema/import.go [hunk 2]
```diff
@@ -21,14 +21,15 @@ package sema
...
 	"github.com/onflow/cadence/runtime/ast"
 	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/common/orderedmap"
 )
 
 // Import
 
 type Import interface {
-	AllValueElements() *StringImportElementOrderedMap
+	AllValueElements() *orderedmap.OrderedMap[string, ImportElement]
```
Comment (SupunS) [inline runtime/sema/import.go:30]: maybe define a type/alias for `*orderedmap.OrderedMap[string, ImportElement]`

File: runtime/sema/type.go [hunk 3]
```diff
@@ -155,7 +156,7 @@ type Type interface {
 	//
 	Unify(
 		other Type,
-		typeParameters *TypeParameterTypeOrderedMap,
+		typeParameters *orderedmap.OrderedMap[*TypeParameter, Type],
```
Comment (SupunS) [inline runtime/sema/type.go:159]: similarly, maybe it's worth defining type aliases for these commonly used map types.

**Review by SupunS — APPROVED** [review]
`+400 −3,235 ` 😍

File: runtime/sema/import.go (hunk 2 above)
Comment (turbolent) [reply runtime/sema/import.go]: Good point! Added in [9045c19](https://github.com/onflow/cadence/pull/1804/commits/9045c1981258031fe0f9946dea9c4963b44d7e9d). That shaved off another ~100 LOC :-)

File: runtime/sema/type.go (hunk 3 above)
Comment (turbolent) [reply runtime/sema/type.go]: Added in [9045c19](https://github.com/onflow/cadence/pull/1804/commits/9045c1981258031fe0f9946dea9c4963b44d7e9d)

**Review by SupunS — APPROVED** [review]

**Comment by SaveTheRbtz** [comment]
Yey!  Once available we could also switch to a generic-based version of containers (e.g. `container/list` is still interface-based)

**Pull request merged**